File: Algo/Algorithm/Questions/Arrays/20.apartmentHunting/main.js
// O(b^2*r) Time | O(b) Space
// function distanceBetween(a, b) {
//   return Math.abs(a - b)
// }

// function getIndxMinValue(array) {
//   let index = 0;
//   let minValue = Infinity;

//   for(let i = 0; i < array.length; i++) {
//     let curr = array[i]

//     if (curr < minValue) {
//       minValue = curr
//       index = i
//     }
//   }

//   return index
// }

// function apartmentHunting(blocks, reqs) {
//   let maxDistanceAtBlock = new Array(blocks.length).fill(-Infinity)

//   for(let i = 0; i < blocks.length; i++) {
//     for(req of reqs) {
//       let closestRequest = Infinity

//       for(let j = 0; j < blocks.length; j++) {

//         if(blocks[j][req]) {
//           closestRequest = Math.min(closestRequest, distanceBetween(i, j))
//         }
//       }

//       maxDistanceAtBlock[i] = Math.max(maxDistanceAtBlock[i], closestRequest)
//     }
//   }

//   return getIndxMinValue(maxDistanceAtBlock)
// }


// O(br) Time | O(br) Space
function distanceBetween(a, b) {
  return Math.abs(a - b)
}

function getMinDistanceFromBlocks(blocks, req) {
  let minDistances = new Array(blocks.length).fill(Infinity)
  let closestReqIndx = Infinity

  for(let i = 0; i < blocks.length; i++) {
    if (blocks[i][req]) {
      closestReqIndx = i
    }
    minDistances[i] = distanceBetween(i, closestReqIndx)
  }

  for(let i = blocks.length - 1; i >= 0; i--) {
    if (blocks[i][req]) {
      closestReqIndx = i
    }
    minDistances[i] = Math.min(minDistances[i], distanceBetween(i, closestReqIndx))
  }

  return minDistances
}

function getMaxDistanceFromBlocks(blocks, minDistanceFromBlocks) {
  let maxDistances = new Array(blocks.length).fill(0)
  for(let i = 0; i < blocks.length; i++) {
    minDistanceFromBlocks.forEach(min => {
      maxDistances[i] = Math.max(min[i], maxDistances[i])
    })
  }

  return maxDistances
}

function getIndxMinValue(array) {
  let index = 0;
  let minValue = Infinity;

  for(let i = 0; i < array.length; i++) {
    let curr = array[i]

    if (curr < minValue) {
      minValue = curr
      index = i
    }
  }

  return index
}

function apartmentHunting(blocks, reqs) {
  const minDistanceFromBlocks = reqs.map(req => getMinDistanceFromBlocks(blocks, req))
  const maxDistanceFromBlocks = getMaxDistanceFromBlocks(blocks, minDistanceFromBlocks)
  return getIndxMinValue(maxDistanceFromBlocks)
}

module.exports = apartmentHunting;

File: Algo/Algorithm/Questions/Arrays/20.apartmentHunting/main.cpp
#include <vector>
#include <unordered_map>
#include <climits>
#include <iostream>
using namespace std;

using namespace std;

// O(br) time | O(br) space - where b is the number of blocks and r is the number of requirements.
vector<int> minDistances(vector<unordered_map<string, bool>> blocks, string req)
{
  vector<int> distances(blocks.size(), 0);
  int closestReq = INT_MAX;
  for (int i = 0; i < blocks.size(); i++)
  {
    if (blocks[i][req])
      closestReq = i;
    distances[i] = abs(i - closestReq);
  }

  closestReq = INT_MAX;
  for (int i = (blocks.size() - 1); i >= 0; i--)
  {
    if (blocks[i][req])
      closestReq = i;
    distances[i] = min(distances[i], abs(closestReq - i));
  }

  return distances;
}

int maxDistances(unordered_map<string, vector<int>> distances, vector<string> reqs, int index)
{
  int myIndex = INT_MIN;
  for (string req : reqs)
    myIndex = max(myIndex, distances[req][index]);
  return myIndex;
}

int bestBlock(vector<int> maxDistanceOfAReq)
{
  int myIndex = 0;
  int myValue = INT_MAX;
  for (int i = 0; i < maxDistanceOfAReq.size(); i++)
  {
    if (maxDistanceOfAReq[i] < myValue)
    {
      myIndex = i;
      myValue = maxDistanceOfAReq[i];
    }
  }
  return myIndex;
}

int apartmentHunting(vector<unordered_map<string, bool>> blocks,
                     vector<string> reqs)
{
  unordered_map<string, vector<int>> minDistanceFromBlock;
  vector<int> maxDistanceOfAReq(blocks.size(), 0);
  for (string req : reqs)
    minDistanceFromBlock[req] = minDistances(blocks, req);

  for (int i = 0; i < blocks.size(); i++)
    maxDistanceOfAReq[i] = maxDistances(minDistanceFromBlock, reqs, i);

  return bestBlock(maxDistanceOfAReq);
}

int main()
{

  vector<unordered_map<string, bool>>
      blocks = {
          {{"gym", false},
           {"school", true},
           {"store", false}},
          {{"gym", true},
           {"school", false},
           {"store", false}},
          {{"gym", true},
           {"school", true},
           {"store", false}},
          {{"gym", false},
           {"school", true},
           {"store", false}},
          {{"gym", false},
           {"school", true},
           {"store", true}}};

  vector<string> reqs = {"gym", "school", "store"};

  cout << apartmentHunting(blocks, reqs) << endl;
  return 0;
}